crates/storage/provider/src/providers/rocksdb/invariants.rs

```diff
@@ -0,0 +1,413 @@
+
+impl RocksDBProvider {
+    /// Checks consistency of `RocksDB` tables against MDBX stage checkpoints.
+    ///
+    /// Returns an unwind target block number if the pipeline needs to unwind to rebuild
+    /// `RocksDB` data. Returns `None` if all invariants pass or if inconsistencies were healed.
+    ///
+    /// # Invariants checked
+    ///
+    /// For `TransactionHashNumbers`:
+    /// - The maximum `TxNumber` value should not exceed what the `TransactionLookup` stage
+    ///   checkpoint indicates has been processed.
+    /// - If `RocksDB` is ahead, excess entries are pruned (healed).
+    /// - If `RocksDB` is behind, an unwind is required.
+    ///
+    /// For `StoragesHistory`:
+    /// - The maximum block number in shards should not exceed the `IndexStorageHistory` stage
+    ///   checkpoint.
+    /// - Similar healing/unwind logic applies.
+    pub fn check_consistency<Provider>(
+        &self,
+        provider: &Provider,
+    ) -> ProviderResult<Option<BlockNumber>>
+    where
+        Provider: DBProvider + StageCheckpointReader + StorageSettingsCache,
+    {
+        let mut unwind_target: Option<BlockNumber> = None;
+
+        // Check TransactionHashNumbers if stored in RocksDB
+        if provider.cached_storage_settings().transaction_hash_numbers_in_rocksdb &&
+            let Some(target) = self.check_transaction_hash_numbers(provider)?
+        {
+            unwind_target = Some(unwind_target.map_or(target, |t| t.min(target)));
+        }
+
+        // Check StoragesHistory if stored in RocksDB
+        if provider.cached_storage_settings().storages_history_in_rocksdb &&
+            let Some(target) = self.check_storages_history(provider)?
+        {
+            unwind_target = Some(unwind_target.map_or(target, |t| t.min(target)));
+        }
+
+        Ok(unwind_target)
+    }
+
+    /// Checks invariants for the `TransactionHashNumbers` table.
+    ///
+    /// Returns a block number to unwind to if `RocksDB` is behind the checkpoint.
+    fn check_transaction_hash_numbers<Provider>(
+        &self,
+        provider: &Provider,
+    ) -> ProviderResult<Option<BlockNumber>>
+    where
+        Provider: DBProvider + StageCheckpointReader,
+    {
+        use reth_db::cursor::DbCursorRO;
+        use reth_db_api::transaction::DbTx;
+
+        // Get the TransactionLookup stage checkpoint
+        let checkpoint = provider
+            .get_stage_checkpoint(StageId::TransactionLookup)?
+            .map(|cp| cp.block_number)
+            .unwrap_or(0);
+
+        // Get the last entry from RocksDB TransactionHashNumbers
+        let rocks_last = self.last::<tables::TransactionHashNumbers>()?;
+
+        match rocks_last {
+            Some((_hash, tx_number)) => {
+                // If checkpoint is 0 but we have data, that's inconsistent
+                if checkpoint == 0 && tx_number > 0 {
+                    // RocksDB has data but stage hasn't run - need to clear
+                    return Ok(Some(0));
```

Comment:
instead of issuing an unwind we could just prune it here

Reply:
yep, done